Interface.py

```diff
@@ -1,72 +1,195 @@
+            return True
 
-        return False
     def cancelled(self):
-        return True
+        r = requests.get(url = self.url_str + '/grab-job-info', json = {'job_id': self.job_id})
+        return r.json()['cancelled']
 
+    #distinct from the done method as this works whether started or not
     def running(self):
-        return True
+        #make sure you check to see if it was started, DO NOT depend on done()
+         r = requests.get(url = self.url_str + '/grab-job-info', json = {'job_id': self.job_id})
+         return r.json()['running']
+
     def done(self):
+         r = requests.get(url = self.url_str + '/grab-job-info', json = {'job_id': self.job_id})
+         if (r.json()['cancelled'] or r.json()['complete']):
+             return True
+         else:
+             return False
+
+    #can't do regular "done" thing, needs to 
+        #raise CancelledError
+        #and also mirror the exception raised by the call
+    def result(self, timeout = None):
+        time_check = True
+        return_value = None
+        timer = 0.0
+        while (time_check):
+            time.sleep(1.0)
+            timer += 1.0
+            r = requests.get(url = self.url_str + '/grab-job-info', json = {'job_id': self.job_id})
+            if (r.json['cancelled']):
+                raise CancelledError
+            if (r.json['exception'] is not None):
+                #develop system to have the same exceptions
+                raise r.json['exception']
+            if (r.json()['complete'] == True):
+                break
+            if (timer > timeout):
+                raise TimeoutError  
         return True
-    def result(self):
-        success = False
 
-        if success:
-            return success
-        else:
-            return success
+    #we are talking about the exceptions in the job itself
     def exception(self, timeout = None):
-        #check for successful run without any exceptions
+        time_check = True
+        return_value = None
+        timer = 0.0
+        while (time_check):
+            time.sleep(1.0)
+            timer += 1.0
+            r = requests.get(url = self.url_str + '/grab-job-info', json = {'job_id': self.job_id})
+            if (r.json['cancelled']):
+                raise CancelledError
+            if (r.json()['complete'] == True):
+                break
+            if (timer > timeout):
+                raise TimeoutError
+        #per project parameters, this will return None if no exception is there
+        return r.json()['exception']
+
+
+    #this just runs a function directly after the completion of the job
+    #can assume these functions are only going to reference the job info itself, 
+        #as they take the future object as the only parameter
+        #they can be as simple as printing the results of the job
+                ##########
+    def add_done_callback(self, fn):

```

Comment:
may be call the call back function `callback`. It is less confusing

Reply:
add_done_callback means to register a function to run after the job finishes. It should not block the client code. I will explian its implementation in person